src/libspu/mpc/swift/conversion.cc

```diff
@@ -0,0 +1,298 @@
+
+  auto decompose_in = NdArrayRef(makeType<AShrTy>(field), {decompose_numel});
+
+  DISPATCH_ALL_FIELDS(field, [&]() {
+    using el_t = ring2k_t;
+    using ashr_t = std::array<el_t, 3>;
+
+    NdArrayView<ashr_t> _in(in);
+    NdArrayView<ashr_t> _decompose_in(decompose_in);
+
+    pforeach(0, numel, [&](int64_t idx) {
+      for (int64_t bit_idx = 0; bit_idx < nbits; bit_idx++) {
+        _decompose_in[idx * nbits + bit_idx][0] =
+            static_cast<el_t>(_in[idx][0] >> bit_idx) & 0x1;
+        _decompose_in[idx * nbits + bit_idx][1] =
+            static_cast<el_t>(_in[idx][1] >> bit_idx) & 0x1;
+        _decompose_in[idx * nbits + bit_idx][2] =
+            static_cast<el_t>(_in[idx][2] >> bit_idx) & 0x1;
+      }
+    });
+  });
+
+  // Bit2A
+  // Step0: joint share alpha_b1, alpha_b2, beta_b
+  // Step1: alpha_b = (alpha_b1 ^ alpha_b2)
+  //                = alpha_b1 + alpha_b2 - 2 * alpha_b1 * alpha_b2
+  // Step2: b = (beta_b ^ alpha_b)
+  //          = beta_b + alpha_b - 2 * beta_b * alpha_b
+
+  NdArrayRef alpha_b1(makeType<AShrTy>(field), {decompose_numel});
+  NdArrayRef alpha_b2(makeType<AShrTy>(field), {decompose_numel});
+  NdArrayRef beta_b(makeType<AShrTy>(field), {decompose_numel});
+
+  if (rank == 0) {
+    alpha_b1 = getFirstShare(decompose_in);
+    alpha_b2 = getSecondShare(decompose_in);
+  }
+  if (rank == 1) {
+    alpha_b1 = getFirstShare(decompose_in);
+    beta_b = getSecondShare(decompose_in);
+  }
+  if (rank == 2) {
+    alpha_b2 = getFirstShare(decompose_in);
+    beta_b = getSecondShare(decompose_in);
+  }
+
+  // P0, P1 joint share alpha_b1
+  alpha_b1 = JointSharing(ctx, alpha_b1, 0, 1, "alpha_b1");
+
+  // P0, P2 joint share alpha_b2
+  alpha_b2 = JointSharing(ctx, alpha_b2, 0, 2, "alpha_b2");
+
+  // P1, P2 joint share beta_b
+  beta_b = JointSharing(ctx, beta_b, 1, 2, "beta_b");
+
+  // Step1: alpha_b = (alpha_b1 ^ alpha_b2)
+  //                = alpha_b1 + alpha_b2 - 2 * alpha_b1 * alpha_b2
+  auto tmp = mult.proc(ctx, alpha_b1, alpha_b2);
+
+  auto alpha_b = add.proc(ctx, alpha_b1, alpha_b2);
+  auto neg_tmp = neg.proc(ctx, tmp);
+  alpha_b = add.proc(ctx, alpha_b, neg_tmp);
+  alpha_b = add.proc(ctx, alpha_b, neg_tmp);
+
+  // Step2: b = (beta_b ^ alpha_b)
+  //          = beta_b + alpha_b - 2 * beta_b * alpha_b
+  tmp = mult.proc(ctx, beta_b, alpha_b);
+  neg_tmp = neg.proc(ctx, tmp);
+  auto decompose_out = add.proc(ctx, alpha_b, beta_b);
+  decompose_out = add.proc(ctx, decompose_out, neg_tmp);
+  decompose_out = add.proc(ctx, decompose_out, neg_tmp);
+
+  DISPATCH_ALL_FIELDS(field, [&]() {
+    using el_t = ring2k_t;
+    using ashr_t = std::array<el_t, 3>;
+
+    NdArrayView<ashr_t> _res(res);
+    NdArrayView<ashr_t> _decompose_out(decompose_out);
+
+    el_t tmp_sum0, tmp_sum1, tmp_sum2;
+
+    pforeach(0, numel, [&](int64_t idx) {
+      // use _decompose_out[i*nbits, (i+1)*nbits) to construct res[i]
+      tmp_sum0 = static_cast<el_t>(0);
```

Comment:
`pforeach`与普通的for-loop不同，会将`numel`个任务切分到若干个线程上，此处引用线程外的`tmp_sum0, tmp_sum1, tmp_sum2`是非常危险的！直接创建临时变量就好，不要复用。
```suggestion
      el_t tmp_sum0 = 0;
```